org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaBuildController.java

```diff
@@ -21,9 +38,122 @@
   /** Executes the user code and sends the output of that code across the established websocket. */
   @MessageMapping(Destinations.EXECUTE_CODE)
   @SendToUser(Destinations.PTP_PREFIX + Destinations.OUTPUT_CHANNEL)
-  public UserProgramOutput execute(UserProgram userProgram, Principal principal) {
-    // Send fake output
-    compileRunService.sendMessages(principal.getName(), "Compiling...");
-    return new UserProgramOutput("Hello World!");
+  public UserProgramOutput execute(UserProgram userProgram, Principal principal) throws Exception {
+    // TODO: CSA-48 Handle more than one file
+    String filename = userProgram.getFileName();
+    // We expect the filename to have no .java suffix, remove it if necessary.
+    if (filename.endsWith(".java")) {
+      userProgram.setFileName(filename.substring(0, filename.indexOf(".java")));
+    }
+    executeHelper(userProgram, principal);
+    return new UserProgramOutput("Done!");
+  }
+
+  public void executeHelper(UserProgram userProgram, Principal principal) {
+    File tempFolder = null;
+    try {
+      tempFolder = Files.createTempDirectory("tmpdir").toFile();
+      DiagnosticCollector<JavaFileObject> diagnostics = new DiagnosticCollector<JavaFileObject>();
+
+      CompilationTask task = getCompilationTask(userProgram, tempFolder, diagnostics);
+      compileRunService.sendMessages(principal.getName(), "Compiling your program...");
+      boolean success = task.call();
+
+      // diagnostics will include any compiler errors
+      for (Diagnostic<? extends JavaFileObject> diagnostic : diagnostics.getDiagnostics()) {
+        compileRunService.sendMessages(principal.getName(), diagnostic.toString());
+      }
+
+      if (success) {
+        // set System.out to be a specific output stream in order to capture output of the
+        // program and send it back to the user
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        System.setOut(out);
+
+        compileRunService.sendMessages(principal.getName(), "Compiled!");
+        compileRunService.sendMessages(principal.getName(), "Running your program...");
+        runClass(tempFolder.toURI().toURL(), userProgram.getFileName(), principal);
+
+        // outputStream should now contain output of userProgram
+        outputStream.flush();
+        String result = outputStream.toString();
```

Comment:
Is it possible to log more often? If I run something like this, the output doesn't show up until the end.
```
while(x < 10) {
  x++;
  Thread.sleep(1000);
  System.out.println("Hello World!");
}
```
Also not something that has to be addressed now (but should probably track).

Reply:
added a [task](https://codedotorg.atlassian.net/browse/CSA-137)!